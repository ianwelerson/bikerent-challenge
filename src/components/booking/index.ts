export { default as BookingAddressMap } from './BookingAddressMap.vue'
export { default as BookingPricing } from './BookingPricing.vue'
export { default as BookingOverview } from './BookingOverview.vue'
export { default as BookingSuccess } from './BookingSuccess.vue'
