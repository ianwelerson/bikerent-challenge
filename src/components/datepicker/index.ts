export { default as Datepicker } from './DatepickerInput.vue'
