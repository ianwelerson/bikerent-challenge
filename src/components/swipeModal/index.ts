export { default as SwipeModal } from './SwipeModal.vue'
