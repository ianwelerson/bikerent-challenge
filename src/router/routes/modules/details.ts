import type { RouteRecordRaw } from 'vue-router'

export default [
  {
    path: '/details/:id',
    name: 'details',
    component: () => import('@/views/BikeDetailsView.vue')
  },
  {
    path: '/details/:id/booking',
    name: 'booking',
    component: () => import('@/views/BikeBookingView.vue')
  }
] as RouteRecordRaw[]
