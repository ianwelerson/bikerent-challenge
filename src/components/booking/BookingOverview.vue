<script lang="ts">
import { defineComponent, type PropType } from 'vue'
import axios from 'axios'

import { CurrencyCode } from '@/core/config'

// import { bike } from '@/core/api'
import type { BikeRentDetails } from '@/core/api/modules/typings/bike'

import { BookingPricing } from '@/components/booking'
import { LoadingSpinner } from '@/components/loading'

export default defineComponent({
  name: 'BookingOverview',
  components: {
    BookingPricing,
    LoadingSpinner
  },
  props: {
    rentDetails: {
      type: Object as PropType<BikeRentDetails>,
      required: true
    }
  },
  emits: ['booked'],
  data: () => ({
    currency: CurrencyCode.EUR,
    isLoading: false,
    error: null as string | null
  }),
  watch: {
    isLoading() {
      this.error = null
    }
  },
  methods: {
    errorHandle(message: string | null) {
      this.error = message
    },
    async handleAddBooking() {
      this.isLoading = true

      try {
        // await bike.rent(this.rentDetails)
        await new Promise((resolve) => {
          setTimeout(() => {
            resolve(null)
          }, 3000)
        })
        this.$emit('booked')
      } catch (error) {
        if (axios.isAxiosError(error)) {
          this.errorHandle(error.response?.data.message)
        }
      } finally {
        this.isLoading = false
      }
    }
  }
})
</script>

<template>
  <div>
    <h3 class="text-base mb-4">Booking Overview</h3>

    <div class="divider" />

    <booking-pricing
      :currency="currency"
      :details="rentDetails"
      class="mb-8"
      @updated="isLoading = false"
      @error="errorHandle"
    />

    <button
      :class="['button button--primary w-full py-5', { 'btn-disabled': isLoading || !!error }]"
      :disabled="isLoading || !!error"
      @click="handleAddBooking"
    >
      <loading-spinner v-if="isLoading" />
      <div v-else-if="!!error" class="error-message">{{ error }}</div>
      <template v-else>Add to booking</template>
    </button>
  </div>
</template>

<style scoped lang="scss">
.error-message {
  color: get-theme-color('error');
}
</style>
