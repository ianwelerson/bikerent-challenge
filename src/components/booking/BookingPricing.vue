<script lang="ts">
import { defineComponent, type PropType } from 'vue'
import axios from 'axios'
import { Icon } from '@/components/icon/'
import { formatCurrency } from '@/core/helpers/currency'
import { bike } from '@/core/api'

import type { CurrencyCode } from '@/core/config'
import type { BikeRentDetails } from '@/core/api/modules/typings/bike'

export default defineComponent({
  name: 'BookingPricing',
  components: { Icon },
  props: {
    currency: {
      type: String as PropType<CurrencyCode>,
      required: true
    },
    details: {
      type: Object as PropType<BikeRentDetails>,
      required: true
    }
  },
  emits: ['loading', 'updated', 'error'],
  data: () => ({
    subtotal: 0,
    serviceFee: 0,
    total: 0
  }),
  watch: {
    details() {
      this.updatePrice()
    }
  },
  mounted() {
    this.updatePrice()
  },
  methods: {
    formatPrice(value: number): String {
      return formatCurrency(value, this.currency)
    },
    async updatePrice() {
      this.$emit('loading')

      try {
        const { rentAmount, fee, totalAmount } = await bike.amount(this.details)

        this.subtotal = rentAmount
        this.serviceFee = fee
        this.total = totalAmount
      } catch (error) {
        if (axios.isAxiosError(error)) {
          this.$emit('error', error.response?.data.message)
        }
      } finally {
        this.$emit('updated')
      }
    }
  }
})
</script>

<template>
  <section class="booking-pricing">
    <div class="grid grid-cols-2 gap-y-3 text-sm">
      <div class="flex items-center">
        Subtotal
        <icon type="solid" size="xs" class="ml-1">circle-info</icon>
      </div>
      <div class="text-right">{{ formatPrice(subtotal) }}</div>

      <div class="flex items-center">
        Service Fee
        <icon type="solid" size="xs" class="ml-1">circle-info</icon>
      </div>
      <div class="text-right">{{ formatPrice(serviceFee) }}</div>

      <div class="flex items-center text-base">Total</div>
      <div class="text-right text-2xl font-bold">{{ formatPrice(total) }}</div>
    </div>
  </section>
</template>

<style lang="scss" scoped>
.booking-pricing {
  position: relative;
}
</style>
