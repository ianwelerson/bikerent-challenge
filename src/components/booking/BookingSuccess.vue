<script lang="ts">
import { defineComponent } from 'vue'

import { Chip } from '@/components/chip'
import { ImageLazy } from '@/components/image'

export default defineComponent({
  name: 'BookingSuccess',
  components: {
    Chip,
    ImageLazy
  },
  props: {
    thumb: {
      type: String,
      required: true
    },
    name: {
      type: String,
      required: true
    },
    type: {
      type: String,
      required: true
    },
    showRedirect: {
      type: Boolean,
      default: false
    }
  }
})
</script>

<template>
  <div class="rent-success">
    <div>
      <h3 class="mb-6 text-2xl font-extrabold">Thank you!</h3>
      <p class="m-0 text-base font-semibold">Your bike is booked.</p>
    </div>
    <div class="mt-6">
      <image-lazy :src="thumb" class="rent-success__img" />
      <div class="mt-4">
        <h2 class="text-lg font-bold">{{ name }}</h2>
        <chip color="secondary" size="sm">{{ type }}</chip>
      </div>
    </div>
    <div v-if="showRedirect" class="mt-5">
      <router-link :to="{ name: 'home' }" class="button button--primary w-full">Go to Home Page</router-link>
    </div>
  </div>
</template>

<style scoped lang="scss">
.rent-success {
  padding: 39px 0;
  @apply .text-center;

  &__img {
    width: 185px;
  }
}
</style>
