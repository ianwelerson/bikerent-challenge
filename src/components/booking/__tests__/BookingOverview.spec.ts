import { beforeEach, describe, expect, it, vi } from 'vitest'
import { mount, flushPromises } from '@vue/test-utils'

import { mockBikeRentDetails, mockedBikeReturnDetails } from '@/mocks'

import { BookingOverview } from '@/components/booking'
import { rent } from '@/core/api/modules/bike'

const mockProps = { rentDetails: mockBikeRentDetails }

vi.mock('@/core/api/modules/bike', () => ({
  rent: vi.fn().mockReturnValue({
    ...mockedBikeReturnDetails
  })
}))

describe('BookingOverview', () => {
  beforeEach(() => {
    vi.clearAllMocks()
  })

  it('renders correctly', () => {
    const wrapper = mount(BookingOverview, { props: { ...mockProps } })

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should be able to rent a bike', async () => {
    const wrapper = mount(BookingOverview, { props: { ...mockProps } })

    await wrapper.find('.button--primary').trigger('click')

    await flushPromises()

    expect(rent).toHaveBeenCalledTimes(1)
    expect(wrapper.emitted().booked).toBeTruthy()
  })
})
