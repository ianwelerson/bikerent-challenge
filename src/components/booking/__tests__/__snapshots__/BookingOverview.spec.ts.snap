// Vitest Snapshot v1

exports[`BookingOverview > renders correctly 1`] = `
"<div data-v-30f699bc=\\"\\">
  <h3 class=\\"text-base mb-4\\" data-v-30f699bc=\\"\\">Booking Overview</h3>
  <div class=\\"divider\\" data-v-30f699bc=\\"\\"></div>
  <section class=\\"booking-pricing mb-8\\" data-v-c7ba2ef4=\\"\\" data-v-30f699bc=\\"\\">
    <div class=\\"grid grid-cols-2 gap-y-3 text-sm\\" data-v-c7ba2ef4=\\"\\">
      <div class=\\"flex items-center\\" data-v-c7ba2ef4=\\"\\"> Subtotal <div class=\\"icon text-xs ml-1\\" data-v-6ea13f32=\\"\\" data-v-c7ba2ef4=\\"\\"><em class=\\"fa-solid fa-circle-info\\" data-v-6ea13f32=\\"\\"></em></div>
      </div>
      <div class=\\"text-right\\" data-v-c7ba2ef4=\\"\\">0 €</div>
      <div class=\\"flex items-center\\" data-v-c7ba2ef4=\\"\\"> Service Fee <div class=\\"icon text-xs ml-1\\" data-v-6ea13f32=\\"\\" data-v-c7ba2ef4=\\"\\"><em class=\\"fa-solid fa-circle-info\\" data-v-6ea13f32=\\"\\"></em></div>
      </div>
      <div class=\\"text-right\\" data-v-c7ba2ef4=\\"\\">0 €</div>
      <div class=\\"flex items-center text-base\\" data-v-c7ba2ef4=\\"\\">Total</div>
      <div class=\\"text-right text-2xl font-bold\\" data-v-c7ba2ef4=\\"\\">0 €</div>
    </div>
  </section><button class=\\"button button--primary w-full py-5\\" data-v-30f699bc=\\"\\">Add to booking</button>
</div>"
`;
